frame/reward-distribution/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+            let (stakers_imbalance, remainder) = remainder.split(total_staker_balance);
+            let (collator_imbalance, remainder) = remainder.split(collator_balance);
+            let (treasury_imbalance, remainder) = remainder.split(total_treasury_balance);
+
+            // TODO: remove this later, useful for testing atm
+            assert!(remainder.peek().is_zero());
+
+            // Payout beneficiaries
+            T::BeneficiaryPayout::treasury(treasury_imbalance);
+            T::BeneficiaryPayout::collators(collator_imbalance);
+            T::BeneficiaryPayout::dapps_staking(stakers_imbalance, dapps_imbalance);
+
+            // TODO: benchmark this code
+            100
+        }
+    }
+}
+
+// TODO (remove this)
+// Alternative to this approach would be to have some as configurable constants.
+// This approach, however, provides more flexibility since we can change them without
+// runtime upgrade.
+
+/// List of parameters used to calculate reward distribution portions for all the beneficiaries.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct RewardDistributionParams {
+    /// Base percentage of reward that goes to treasury
+    #[codec(compact)]
+    pub base_treasury_percent: Perbill,
+    /// Base percentage of reward that goes to stakers
+    #[codec(compact)]
+    pub base_staker_percent: Perbill,
+    /// Percentage of rewards that goes to dApps
+    #[codec(compact)]
+    pub dapps_percent: Perbill,
+    /// Percentage of reward that goes to collators
+    #[codec(compact)]
+    pub collators_percent: Perbill,
+    /// Adjustable reward percentage that either goes to treasury or to stakers
+    #[codec(compact)]
+    pub adjustable_percent: Perbill,
+    /// Target dapps-staking TVL percentage at which adjustable inflation towards stakers becomes saturated
+    #[codec(compact)]
+    pub ideal_dapps_staking_tvl: Perbill,
+}
+
+impl Default for RewardDistributionParams {
+    fn default() -> Self {
+        RewardDistributionParams {
+            base_treasury_percent: Perbill::from_percent(40),
+            base_staker_percent: Perbill::from_percent(25),
+            dapps_percent: Perbill::from_percent(25),
+            collators_percent: Perbill::from_percent(10),
+            adjustable_percent: Zero::zero(),
+            ideal_dapps_staking_tvl: Zero::zero(),
+        }
+    }
+}
+
+impl RewardDistributionParams {
+    /// `true` if sum of all percentages is `one whole`, `false` otherwise.
+    fn verify(&self) -> bool {
+        // TODO: probably can be implemented in a cleaner way
+        let sum = self
+            .base_treasury_percent
+            .saturating_add(self.base_staker_percent)
+            .saturating_add(self.dapps_percent)
+            .saturating_add(self.collators_percent);
+
+        (Perbill::one() - sum) == self.adjustable_percent
+    }
+}
+
+pub trait TvlProvider<Balance> {
```

Comment:
Please add trait description that explains the cause of introducing this trait.

Reply:
Sure, I haven't really documented the pallet in details since this was just supposed to be a PR to give an overview of the idea.

In the end, pallet will have proper documentation everywhere.